stdlib/public/Concurrency/Actor.cpp

```diff
@@ -935,356 +1034,370 @@ static void traceJobQueue(DefaultActorImpl *actor, Job *first) {
-  runJobInEstablishedExecutorContext(job);
+    if (!firstJob) {
+      // Nothing to drain, short circuit
+      SWIFT_TASK_DEBUG_LOG("No jobs to drain on actor %p", this);
+      return NULL;
+    }
 
-  concurrency::trace::job_run_end(job, &executor, traceHandle);
-  trackingInfo.leave();
+    auto newState = oldState;
+    // Dequeue the first job and set up a new head
+    newState = newState.withFirstJob(getNextJobInQueue(firstJob));
+    if (CurrentState.compare_exchange_weak(oldState, newState,
+                            /* sucess */ std::memory_order_release,
+                            /* failure */ std::memory_order_acquire)) {
+      SWIFT_TASK_DEBUG_LOG("Drained first job %p from actor %p", firstJob, this);
+      traceActorStateTransition(this, oldState, newState);
+      return firstJob;
+    }
 
-  // Give up the current executor if this is a switching context
-  // (which, remember, only happens if we started out on a generic
-  // executor) and we've switched to a default actor.
-  auto currentExecutor = trackingInfo.getActiveExecutor();
-  if (trackingInfo.allowsSwitching() && currentExecutor.isDefaultActor()) {
-    // Use an underestimated priority; if this means we create an
-    // extra processing job in some cases, that's probably okay.
-    auto runner = RunningJobInfo::forOther(JobPriority(0));
-    asImpl(currentExecutor.getDefaultActor())->giveUpThread(runner);
+    // There were new item concurrently added to the queue so our "dequeue"
+    // didn't work. Preprocess stuff starting from the top
+    firstJob = preprocessQueue(oldState.getFirstJob());
+    traceJobQueue(this, firstJob);
   }
 }
 
-/// The primary function for processing an actor on a thread.  Start
-/// processing the given default actor as the active default actor on
-/// the current thread, and keep processing whatever actor we're
-/// running when code returns back to us until we're not processing
-/// any actors anymore.
-///
-/// \param currentActor is expected to be passed in as retained to ensure that
-///                     the actor lives for the duration of job execution.
-///                     Note that this may conflict with the retain/release
-///                     design in the DefaultActorImpl, but it does fix bugs!
-SWIFT_CC(swiftasync)
-static void processDefaultActor(DefaultActorImpl *currentActor,
-                                RunningJobInfo runner) {
-  SWIFT_TASK_DEBUG_LOG("processDefaultActor %p", currentActor);
-  DefaultActorImpl *actor = currentActor;
-
-  // If we actually have work to do, we'll need to set up tracking info.
-  // Optimistically assume that we will; the alternative (an override job
-  // took over the actor first) is rare.
+// Called from processing jobs which are created to drain an actor. We need to
+// reason about this function together with swift_task_switch because threads
+// can switch from "following actors" to "following tasks" and vice versa.
```

Comment:
Ow, my brain.

Reply:
Yea I don't know if that's an easy way to think about it but it's what feels natural to me. We ask for and get a thread to drain an actor but once it starts running a task from the actor, it follows along with the task's execution even as it hops off the actor and doesn't necessarily continue draining the actor.

Thread following task vs thread following actor is therefore how I think about it. I agree it's mind-bending